[clang][LLVM Demangler] Add a diagnostic that validates that all mang…

VitaNuo opened the pull request:
…led names produced by `clang` can be demangled by LLVM demangler.

Introduce the above diagnostic behind the `-fno-demangling-failures` flag to prevent unintended breakages.

erichkeane commented on clang/include/clang/Driver/Options.td:
```diff
@@ -1967,6 +1967,10 @@ def fclang_abi_compat_EQ : Joined<["-"], "fclang-abi-compat=">, Group<f_clang_Gr
   Visibility<[ClangOption, CC1Option]>,
   MetaVarName<"<version>">, Values<"<major>.<minor>,latest">,
   HelpText<"Attempt to match the ABI of Clang <version>">;
+def fno_demangling_failures: Flag<["-"], "fno-demangling-failures">, Group<f_clang_Group>,
```
I don't recall how it works, but I think we want an on & off flag type, not just the off.

VitaNuo replied:
Yeah this flag turns on an assertion that validates that all the manglings produced by clang can be demangled. In this sense it on & off (off by default). Let me know if I misunderstand something.

erichkeane replied:
In that we need a 'turn it back on' flag as well, so we need `-fdemangling-failures`.

VitaNuo replied:
This has turned into `-fdiagnostics-demangler-failures` now.

erichkeane commented on clang/include/clang/Basic/CodeGenOptions.def:
```diff
@@ -462,6 +462,9 @@ ENUM_CODEGENOPT(ZeroCallUsedRegs, llvm::ZeroCallUsedRegs::ZeroCallUsedRegsKind,
 CODEGENOPT(CtorDtorReturnThis, 1, 0)
 
+/// Whether to validate if a produced mangled name can be demangled with LLVM demangler.
+CODEGENOPT(NoDemanglingFailures, 1, 0)
```
This is probably inverted.

VitaNuo replied:
The macro signature is `CODEGENOPT(Name, Bits, Default)`, so I assume you're referring to the `Default` as being inverted (the bit count is 1, since this is a flag).

IIUC `0` means that the flag is off (= don't apply the assertion), `1` means the flag is on (= apply the assertion and validate that mangled names can be demangled). We don't want to set the flag to on by default, so `0` seems right.

Could you explain why you believe this inverted? I'm probably missing something.

erichkeane replied:
I mean the name.  We shouldn't have a true/false flag where the 'true' isn't "turn this thing on".  So this should be `DemanglingFailures`.

VitaNuo replied:
Thank you, I've changed it to `DemanglingFailures`, so that `true` means "turn the diagnostic on".

VitaNuo replied:
This has turned into `DiagnosticsDemanglerFailures` now. Resolving this thread to avoid duplicate discussions.

erichkeane replied:
I think `Diagnose` vs `Diagnostics`, since the point is to enable `diagnose`-ing of the feature makes more sense.

erichkeane reviewed (commented):
Since this is being done as an assert, this is something that really only applies during debug mode.  I dont think this compiler flag makes sense because of it.

IMO, this should either be a preprocessor macro to enable (so it is a build config), or a diagnostic that happens all the time (perhaps whose default changes?).

VitaNuo commented:
> Since this is being done as an assert, this is something that really only applies during debug mode. I dont think this compiler flag makes sense because of it.

The reason I am adding a flag on top of an assert is that there are unfortunately still many mangled names produced by `clang` that cannot be demangled. E.g., many of the existing codegen lit tests (`clang/test/CodeGenCXX/*`) fail in assertion-enabled builds without the flag. This is potentially disruptive to development workflows. The flag can be removed once the LLVM demangler can handle all the mangled names `clang` produces.

ilya-biryukov commented:
Sorry for jumping in late.

I think Erich is on point that having a flag that controls an assertion is a bit of a red flag as we are mixing build configuration and runtime configuration.

It is at least unusual and may cause confusion.

After thinking about this a bit more, should we maybe go all-in on one of the approaches?
- either use build configuration: add a new build flag that controls this assertion. Only assert when assertions are enabled and the new build flag is defined.
- or use runtime configure always, e.g. add diagnostics for names that can't be demangled. It should make finding those issues much easier (one can run the compiler to produce a list of names that can't be demangled with locations pointing at code that helps to identify where those names come from).


I would probably vouch for the second approach. The only downside that I see is that we have to explore the warning flag to users (right?) and this is something that should be internal (-cc1).

What do people think of that proposal?

erichkeane commented:
> Sorry for jumping in late.
> 
> I think Erich is on point that having a flag that controls an assertion is a bit of a red flag as we are mixing build configuration and runtime configuration.
> 
> It is at least unusual and may cause confusion.
> 
> After thinking about this a bit more, should we maybe go all-in on one of the approaches?
> 
>     * either use build configuration: add a new build flag that controls this assertion. Only assert when assertions are enabled and the new build flag is defined.
> 
>     * or use runtime configuration always, e.g. add diagnostics for names that can't be demangled. It should make finding those issues much easier (one can run the compiler to produce a list of names that can't be demangled with locations pointing at code that helps to identify where those names come from).
> 
> 
> I would probably vouch for the second approach. The only downside that I see is that we have to expose the warning flag to users (right?) and this is something that should be internal (-cc1).
> 
> What do people think of that proposal?

I agree completely.  I have a preference for the second thing as well.  We can have this be a warning-as-default-error type thing, which allows us to disable this with a flag, but is still an error.  I MIGHT suggest using the functionality to see if that diagnostic is on before doing the test (due to how much additional work there is here), but IMO, that is a much better way forward.

ilya-biryukov commented:
> I agree completely. I have a preference for the second thing as well. We can have this be a warning-as-default-error type thing, which allows us to disable this with a flag, but is still an error. I MIGHT suggest using the functionality to see if that diagnostic is on before doing the test (due to how much additional work there is here), but IMO, that is a much better way forward.

👍  I guess that's a good signal that this is the way to go. AFAIK, @VitaNuo has already started exploring this direction.
Wrt to enabling this warning, I think we should probably start by having this warning disabled because there seems to be too many failures, but we could gradually roll it out:
- start with a warning that's disabled,
- collect issues notable in LLVM tests (there's plenty AFAIU) and also from real projects (we can run with the warning internally over Google's codebase to get a representative sample),
- fix all demangling issues present in the tests and the most common ones from other projects,
- enabling the flag by default.

@erichkeane how do you feel about having it enabled given that this does not really block users, but merely points at demangling issues in LLVM's demangler (that they don't necessarily even use). It feels like this should be internal, but then we won't get any bug reports. I feel like we should at least fix the majority of the issues before enabling it (hence the rough roadmap above).

erichkeane commented:
> > I agree completely. I have a preference for the second thing as well. We can have this be a warning-as-default-error type thing, which allows us to disable this with a flag, but is still an error. I MIGHT suggest using the functionality to see if that diagnostic is on before doing the test (due to how much additional work there is here), but IMO, that is a much better way forward.
> 
> 👍 I guess that's a good signal that this is the way to go. AFAIK, @VitaNuo has already started exploring this direction. Wrt to enabling this warning, I think we should probably start by having this warning disabled because there seems to be too many failures, but we could gradually roll it out:
> 
>     * start with a warning that's disabled by default,
> 
>     * collect issues that we hit in LLVM tests (there's plenty AFAIU) and also from real projects (we can run with the warning internally over Google's codebase to get a representative sample),
> 
>     * fix all demangling issues present in the tests and the most common ones from other projects,
> 
>     * enable the flag by default.
> 
> 
> @erichkeane how do you feel about having it enabled given that this does not really block users, but merely points at demangling issues in LLVM's demangler (that they don't necessarily even use). It feels like this should be internal, but then we won't get any bug reports. I feel like we should at least fix the majority of the issues before enabling it (hence the rough roadmap above).

I definitely have some heartburn about the enabling-of-flag by default, I would expect these diagnostics to be something actionable by our users, but I also definitely see the wish to get as many people testing this to improve the llvm demangler as possible.  I pinged a few others to comment here, but I'd have to be reasonably sure this is a 'no one will ever see this' kind of thing, and if they DO, have a good idea of what to do.  

I'm on the fence, but if we were to, I have some thoughts:
1- Can we test this on the top X github C++ projects as well?  Perhaps at LEAST Boost.
2- Can we ensure that anyone who searches for the diagnostic has a way of VERY QUICKLY seeing how to report it?  So it has to be VERY well documented.
3- Have the diagnostic make it clear that this isn't a "something wrong with your code, but with our demangler"?
4- Instead of being a warning, how about making it a 'remark'?

VitaNuo commented:
Totally agree to 2 to 4.

Regarding the testing and => default enablement, I think we'll have to first see how many failures we'll see on LLVM + Google internal codebase. It's hard to tell at this point how many of those there will be, and so how feasible it is (or how soon it is feasible) to enable the warning by default.

AaronBallman commented:
> > I agree completely. I have a preference for the second thing as well. We can have this be a warning-as-default-error type thing, which allows us to disable this with a flag, but is still an error. I MIGHT suggest using the functionality to see if that diagnostic is on before doing the test (due to how much additional work there is here), but IMO, that is a much better way forward.
> 
> 👍 I guess that's a good signal that this is the way to go. AFAIK, @VitaNuo has already started exploring this direction. Wrt to enabling this warning, I think we should probably start by having this warning disabled because there seems to be too many failures, but we could gradually roll it out:
> 
>     * start with a warning that's disabled by default,
> 
>     * collect issues that we hit in LLVM tests (there's plenty AFAIU) and also from real projects (we can run with the warning internally over Google's codebase to get a representative sample),
> 
>     * fix all demangling issues present in the tests and the most common ones from other projects,
> 
>     * enable the flag by default.
> 
> 
> @erichkeane how do you feel about having it enabled given that this does not really block users, but merely points at demangling issues in LLVM's demangler (that they don't necessarily even use). It feels like this should be internal, but then we won't get any bug reports. I feel like we should at least fix the majority of the issues before enabling it (hence the rough roadmap above).

I am not super comfortable with enabling the diagnostic by default because it's not something the user can do anything about aside from disable the diagnostic (which means they'll report one issue and probably never report another again because they disabled the warning due to chattiness) and I don't think we want to train users to ignore diagnostics or force them to turn off low-value ones. However, what about this as a compromise: we enable the diagnostic without a diagnostic group (so it cannot be disabled) in all builds up to the release candidate, and then we fully disable it for the actual releases? This way, early adopters can help us find issues, but we're not inflicting any pain on the general public once we release.

(FWIW, I think this is something that should have an RFC for wider community buy-in given the effects.)

VitaNuo committed:
[clang][LLVM Demangler] Add a diagnostic that validates that all mangled names produced by `clang` can be demangled by LLVM demangler.

Introduce the above warning behind the `-fdemangling-failures` flag, since:

a. the diagnostic cannot be fixed by an application developer.
b. the diagnostic is expected to be quite chatty.

ilya-biryukov commented:
> I am not super comfortable with enabling the diagnostic by default because it's not something the user can do anything about aside from disable the diagnostic (which means they'll report one issue and probably never report another again because they disabled the warning due to chattiness) and I don't think we want to train users to ignore diagnostics or force them to turn off low-value ones. However, what about this as a compromise: we enable the diagnostic without a diagnostic group (so it cannot be disabled) in all builds up to the release candidate, and then we fully disable it for the actual releases? This way, early adopters can help us find issues, but we're not inflicting any pain on the general public once we release.
> 
> (FWIW, I think this is something that should have an RFC for wider community buy-in given the effects.)

We would need a way to enable or disable this diagnostic internally, ideally via a runtime switch, so that we don't need to ship a new compiler every time a user hits this warning. I think the same also holds for users that use RC: we want them to report the bug, but they also need to make progress on their testing of RC/whatever else they're doing with it.
The only downside of having a diagnostic group that I see is that it's public and shows up in documentation everywhere. However, given the amount of warnings we have, maybe that's not even a big deal.

I think the way out of it is to enable this diagnostic only after we make sure it is not chatty. So we do a gradual rollout:
- add a warning, disabled by default,
- collect cases where we cannot demangle in important projects by enabling the diagnostic for a single run,
- fix those problems,
- enable the warning by default up to RC builds as proposed,
- collect and fix any problems that happen from RC,
- enable the warning by default in release because it's not chatty anymore.
- (if the demangler is really stable) remove the group so that the diagnostic cannot be disabled.

Hopefully by the last step, we will be discovering almost all names that we cannot demangle during Clang development and users will almost never see it.
It it's rare, I hope this would motivate the users to disable it less (although some still will). If we really want to fight that, we can actually remove the group and force it to be enabled by default.

+1 to having this discussion in the RFC

VitaNuo commented:
> I am not super comfortable with enabling the diagnostic by default because it's not something the user can do anything about aside from disable the diagnostic (which means they'll report one issue and probably never report another again because they disabled the warning due to chattiness) and I don't think we want to train users to ignore diagnostics or force them to turn off low-value ones. However, what about this as a compromise: we enable the diagnostic without a diagnostic group (so it cannot be disabled) in all builds up to the release candidate, and then we fully disable it for the actual releases? This way, early adopters can help us find issues, but we're not inflicting any pain on the general public once we release.

> (FWIW, I think this is something that should have an RFC for wider community buy-in given the effects.)

Sounds good, I'll try to summarize this discussion as an RFC.

> collect issues that we hit in LLVM tests (there's plenty AFAIU) and also from real projects (we can run with the warning internally over Google's codebase to get a representative sample).

It might be a good idea to do this first, in order to collect more data for an RFC and the corresponding discussion.

VitaNuo committed:
[clang][LLVM Demangler] Introduce support for the `-fdemangling-failures` flag to `clang` driver.

erichkeane commented on clang/include/clang/Basic/CodeGenOptions.def:
```diff
@@ -462,6 +462,9 @@ ENUM_CODEGENOPT(ZeroCallUsedRegs, llvm::ZeroCallUsedRegs::ZeroCallUsedRegsKind,
 CODEGENOPT(CtorDtorReturnThis, 1, 0)
 
+/// Whether to issue a diagnostic if a produced mangled name can not be demangled with the LLVM demangler.
+CODEGENOPT(DemanglingFailures, 1, 0)
```
```suggestion
CODEGENOPT(DiagDemanglerFailures, 1, 0)
```

VitaNuo replied:
Thanks, mentioning diagnostics makes sense.

erichkeane commented on clang/include/clang/Basic/DiagnosticFrontendKinds.td:
```diff
@@ -384,4 +384,8 @@ def warn_hlsl_langstd_minimal :
+
+def warn_name_cannot_be_demangled : Warning<
+  "cannot demangle the name '%0'">,
+  InGroup<CXX20Compat>;
```
Group is wrong here, we likely want to give it its own group for the diagnostic.

VitaNuo replied:
Added a new `Demangler` group.

erichkeane commented on clang/lib/CodeGen/CodeGenModule.cpp:
```diff
@@ -41,7 +41,9 @@
 #include "clang/Basic/CharInfo.h"
 #include "clang/Basic/CodeGenOptions.h"
 #include "clang/Basic/Diagnostic.h"
+#include "clang/Basic/DiagnosticFrontend.h"
```
This is a LOT of include additions.  Could we instead create a function "elsewhere" (new .cpp file perhaps?) that does all the work, perhaps just `CodeGenModule::DiagDemanglerFailures`?

VitaNuo replied:
Sorry, most of those are a result of multiple iterations of this patch. 
The only necessary one is `llvm/Demangle/Demangle.h`.
Is that all right to add one include or should I still extract the code? 
Removing all the other includes for now.

erichkeane replied:
Ok,just the 1 is fine then.

VitaNuo replied:
Thanks!

erichkeane reviewed (commented):
Close for me, jsut a couple of small requests.

Thanks for your patience on this response, I've been at LLVMDev all week.

VitaNuo committed:
[clang][LLVM Demangler] Address review comments. Rename the flag that controls the demangler diagnostic. Reduce the new include count in CodeGenModule.cpp. Add a new diagnostic group for demangler failures.

VitaNuo commented on clang/include/clang/Driver/Options.td:
```diff
@@ -1967,6 +1967,10 @@ def fclang_abi_compat_EQ : Joined<["-"], "fclang-abi-compat=">, Group<f_clang_Gr
   Visibility<[ClangOption, CC1Option]>,
   MetaVarName<"<version>">, Values<"<major>.<minor>,latest">,
   HelpText<"Attempt to match the ABI of Clang <version>">;
+def fdemangling_failures: Flag<["-"], "fdemangling-failures">, Group<f_clang_Group>,
```
Thank you, good point. I've looked through the existing flags, there are multiple ones that start with `fdiagnostics_XXX`, so I have renamed the flag to `fdiagnostics_demangler_failures`. 
Let me know if this works.

VitaNuo commented:
> Close for me, jsut a couple of small requests.
> 
> Thanks for your patience on this response, I've been at LLVMDev all week.

Thanks Erich for the approval. 
I will publish an RFC for this in the coming days as well, still gathering some data from test runs.

VitaNuo commented:
The RFC is published https://discourse.llvm.org/t/rfc-clang-diagnostic-for-demangling-failures/82835.

erichkeane commented:
> The RFC is published https://discourse.llvm.org/t/rfc-clang-diagnostic-for-demangling-failures/82835.

Thanks!  I thought i commented elsewhere, but I think `diagnose` vs `diagnostics` for the flags/etc makes more sense:
`fdiagnostics_demangler_failures` should be `-fdiagnose_demangler_failures`, it is the correct language-tense.

Else, the patch is good delta what happens with the RFC.  Please make sure to wait until Aaron calls the RFC.

VitaNuo commented:
> Thanks! I thought i commented elsewhere, but I think `diagnose` vs `diagnostics` for the flags/etc makes more sense: `fdiagnostics_demangler_failures` should be `-fdiagnose_demangler_failures`, it is the correct language-tense.

Thanks! I think it was `demangling` vs. `demangler`, and now it's `diagnostics` vs. `diagnose` :) I'll incorporate that.

> Else, the patch is good delta what happens with the RFC. Please make sure to wait until Aaron calls the RFC.

Ofc. I am also OOO next week.

VitaNuo commented:
After some internal discussions, we have decided to put this work on hold.
It still remains an important problem. I have posted some stats for various open source projects in the [RFC](https://discourse.llvm.org/t/rfc-clang-diagnostic-for-demangling-failures/82835). Internally, we have around 177k violations (mangled names that can’t be demangled by the LLVM demangler).

In case someone in the community would like to pick up the effort, this PR has the code for the diagnostic. I will also [add a comment](https://github.com/llvm/llvm-project/pull/117346) to the codegen module mentioning the invariant that should hold in the future.
